utils.go

```diff
@@ -111,3 +111,29 @@ func (l *limitedReadCloser) Read(buf []byte) (int, error) {
 func (l *limitedReadCloser) Close() error {
 	return l.rc.Close()
 }
+
+func GetContentLength(reader io.Reader) (int64, error) {
+	contentLength := int64(0)
+	switch reader := reader.(type) {
+	case *bytes.Buffer:
+		contentLength = int64(reader.Len())
+	case *bytes.Reader:
+		contentLength = int64(reader.Len())
+	case *strings.Reader:
+		contentLength = int64(reader.Len())
+	case io.Seeker:
+		pos, err := reader.Seek(0, io.SeekEnd)
```

Comment:
this is probably slightly incorrect as the reader could be at a position other than zero:

I think you need to:

1. `startPos, err = reader.Seek(0, io.SeekCurrent)` to determine the current position
2. `totalLength, err = reader.Seek(0, io.SeekEnd)` to determine the length entire stream
3. `reader.Seek(0, startPos)` to restore the original starting position.
4. `return totalLength - startPos` - this is how many bytes are remaining